### PR TITLE
Break interface file into multiple files

### PR DESCRIPTION
**Pull request by varunagrawal**
Big PR alert!

This PR allows breaking the GTSAM wrapper into multiple smaller files. The benefits are multifold:

- Parallel compilation for improved efficiency.
- Takes care of the OOM issue in the CI.
- Smaller files are easier to debug and update.
- Minimal syntax changes for the CMake file.

I have tested this for both Python and Matlab and they work great.

### CODE REVIEW

File: gtsam/base/base.i
Comment (ProfFan) [inline gtsam/base/base.i:113]: newline

File: gtsam/linear/linear.i
Comment (ProfFan) [inline gtsam/linear/linear.i:653]: newline

File: python/gtsam/preamble/linear.h
Comment (ProfFan) [inline python/gtsam/preamble/linear.h:12]: newline

File: python/gtsam/preamble/nonlinear.h
Comment (ProfFan) [inline python/gtsam/preamble/nonlinear.h:12]: newline

File: python/gtsam/preamble/sam.h
Comment (ProfFan) [inline python/gtsam/preamble/sam.h:12]: newline

File: python/gtsam/preamble/sfm.h
Comment (ProfFan) [inline python/gtsam/preamble/sfm.h:12]: newline

File: python/gtsam/preamble/symbolic.h
Comment (ProfFan) [inline python/gtsam/preamble/symbolic.h:12]: newline

File: python/gtsam/specializations/linear.h
Comment (ProfFan) [inline python/gtsam/specializations/linear.h:12]: newline

File: python/gtsam/specializations/navigation.h
Comment (ProfFan) [inline python/gtsam/specializations/navigation.h:12]: newline

File: python/gtsam/specializations/nonlinear.h
Comment (ProfFan) [inline python/gtsam/specializations/nonlinear.h:12]: newline

File: python/gtsam/specializations/sam.h
Comment (ProfFan) [inline python/gtsam/specializations/sam.h:12]: newline

File: python/gtsam/specializations/symbolic.h
Comment (ProfFan) [inline python/gtsam/specializations/symbolic.h:12]: newline

**Review by ProfFan — APPROVED** [review]
LGTM :rocket:

**Review by ProfFan — CHANGES_REQUESTED** [review]
LGTM :rocket: 

But we need to let the users check first, maybe GTSFM @johnwlambert ?

**Comment by varunagrawal** [comment]
All checks (including previous OOM ones) pass!

**Comment by johnwlambert** [comment]
Really nice work @varunagrawal. Thanks for working on this.

Is there any sort of quantitative comparison you could potentially add if you have time? Perhaps Python wrapper compilation time on this branch vs. on master? And maybe there is some way to quantitatively measure the RAM usage with and without?

**Comment by varunagrawal** [comment]
@johnwlambert I have not been able to find a way to measure RAM usage via regular profiling tools, if you have any suggestions, that would be great! The biggest evidence IMO is that all the previously commented out CI cases now pass repeatedly without issues.

As for build time, I am looking into that now but I am not sure if it really makes a difference since builds are done occasionally (unlike running code or scripts) and we can always run nightly builds which are asynchronous to developer time.

**Comment by varunagrawal** [comment]
Using [ClangBuildAnalyzer](https://github.com/aras-p/ClangBuildAnalyzer), I get these metrics.

### Previous

Compilation (173 times):
  - Parsing (frontend):         1033.3 s
  - Codegen & opts (backend):    452.1 s

### Current

Compilation (182 times):
  - Parsing (frontend):         1072.1 s
  - Codegen & opts (backend):    457.2 s


The impact is a difference of ~43 seconds, which isn't bad at all. Then again, if we no longer have OOM errors, we can do nightly builds of both `libgtsam` as well as the python packages, so you can simply do `pip install gtsam-<hash>`.

**Comment by ProfFan** [comment]
Note now the wheels can be built without a custom runner. Amazing!

**Comment by varunagrawal** [comment]
@ProfFan can you explain that last comment again? I assumed the custom runner was the manylinux docker image. Are you referring to a runner as in a script?

**Comment by ProfFan** [comment]
Because of the OOM in GitHub hosted runners, I always run them with a high memory instance allocated in my Google Cloud.

**Comment by varunagrawal** [comment]
@ProfFan I see, that's amazing news!!!

**Comment by ProfFan** [comment]
Also, I tested the wheels and they work fine on my application. @johnwlambert  Could you try the wheels I listed with GTSFM?

**Comment by varunagrawal** [comment]
This also means we can now add an action to the manylinux repo CI to publish the nightly build via a scheduled build. :sunglasses:

**Review by johnwlambert — APPROVED** [review]

**Comment by johnwlambert** [comment]
@ProfFan Thanks for the wheels. I tried them, all the gtsam functionality works fine.

LGTM.

File: wrap/gtwrap/matlab_wrapper/mixins.py
Comment (johnwlambert) [inline wrap/gtwrap/matlab_wrapper/mixins.py:26]: are we missing type hints in this file?

File: wrap/gtwrap/matlab_wrapper/mixins.py
Comment (johnwlambert) [inline wrap/gtwrap/matlab_wrapper/mixins.py:1]: maybe we could define how we are using the term "mix-in" here?

File: wrap/gtwrap/pybind_wrapper.py
Comment (johnwlambert) [inline wrap/gtwrap/pybind_wrapper.py:624]: are we missing type hints here?

File: wrap/tests/test_pybind_wrapper.py
Comment (johnwlambert) [inline wrap/tests/test_pybind_wrapper.py:34]: i think we're missing type hints here

File: wrap/gtwrap/matlab_wrapper/mixins.py
Comment (varunagrawal) [reply wrap/gtwrap/matlab_wrapper/mixins.py:26]: Ah shoot. Updating the matlab wrapper was such an onerous task that I didn't think to add type-hints. That can be a future PR since it is not quite related to this functionality.

File: wrap/gtwrap/pybind_wrapper.py
Comment (varunagrawal) [reply wrap/gtwrap/pybind_wrapper.py:624]: Same as above. Next PR. :)

File: wrap/tests/test_pybind_wrapper.py
Comment (varunagrawal) [reply wrap/tests/test_pybind_wrapper.py:34]: Same as above. Next PR. :)

File: wrap/gtwrap/matlab_wrapper/mixins.py
Comment (varunagrawal) [reply wrap/gtwrap/matlab_wrapper/mixins.py:1]: Mixins are a standard programming feature (we call them traits in C++). As such, it should be self-explanatory.

https://en.wikipedia.org/wiki/Mixin
https://stackoverflow.com/questions/533631/what-is-a-mixin-and-why-are-they-useful

**Review by ProfFan — APPROVED** [review]
:green_circle: Greenlight :)

Ready to takeoff :rocket:

**Comment by varunagrawal** [comment]
Currently #811 and #817 touch the interface file. I'll wait for them to be merged and then update this PR with the changes before merging.

**Comment by ProfFan** [comment]
@varunagrawal Feel free to merge :)

**Comment by varunagrawal** [comment]
Updates made! :tada: 

Will merge once CI passes.

File: gtsam/linear/Coreset.h
Comment (ProfFan) [inline gtsam/linear/Coreset.h:1]: What is this file?

**Review by ProfFan — CHANGES_REQUESTED** [review]
It appears that there are some mis-added files?

File: gtsam/linear/Coreset.h
Comment (varunagrawal) [reply gtsam/linear/Coreset.h]: Ah good catch. This was my bad.

**Review by ProfFan — APPROVED** [review]

**Comment by ProfFan** [comment]
@varunagrawal GO for merge :)

**Comment by dellaert** [comment]
Thanks @varunagrawal and all reviewers !!! This is awesome!

**Pull request merged**
